pkg/apis/pmemcsi/v1alpha1/deployment_types_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2019  Intel Corporation.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package v1alpha1_test
+
+import (
+	"testing"
+
+	"github.com/intel/pmem-csi/pkg/apis"
+	api "github.com/intel/pmem-csi/pkg/apis/pmemcsi/v1alpha1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+func TestDeploymentType(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	err := apis.AddToScheme(scheme.Scheme)
+	Expect(err).Should(BeNil(), "Failed to add api schema")
+
+	RunSpecs(t, "PMEM Operator API test suite")
+}
+
+var _ = Describe("Operator", func() {
+
+	BeforeEach(func() {
+	})
+
+	Context("API", func() {
```

Comment:
Is this where we can verify that invalid values are rejected? If yes, please add such tests.

Reply:
The validation part is not handled inside our code, it is done by Kubernetes(api-server?) as per CRD json schema we provide. I will check if we can add such tests.

Reply:
`deployment_types.go` doesn't seem to enough information information for that kind of validation. For example:
```
// DeviceMode to use to manage PMEM devices. One of lvm, direct
DeviceMode DeviceMode `json:"deviceMode,omitempty"`
```

This "one of lvm, direct" is a restriction that we should enforce via validation, if possible.
`LogLevel` is the (only?) other field where validation seems possible at creation time. 

Everything else is something that the operator must validate because the semantic is too complex. But that also needs tests.

Reply:
I think it's okay to add a TODO item about "input validation" for now, i.e. solve this after merging the PR into the "operator" branch (but before merging into "devel").

Reply:
> For example:
> 
> ```
> // DeviceMode to use to manage PMEM devices. One of lvm, direct
> DeviceMode DeviceMode `json:"deviceMode,omitempty"`
> ```
> 
> This "one of lvm, direct" is a restriction that we should enforce via validation,

This supposed to be enforced by the schema, as we defined it [as enum]( https://github.com/intel/pmem-csi/pull/525/files/81e8d693c9dd88c8802e582461a1f546111d6484#diff-6dbccdd31f0d920af3f26b2bbb144b09R303-R310)

Reply:
Added a TODO for this.

Reply:
> This supposed to be enforced by the schema, as we defined it as [enum](https://github.com/intel/pmem-csi/pull/525/files/81e8d693c9dd88c8802e582461a1f546111d6484#diff-6dbccdd31f0d920af3f26b2bbb144b09R303-R310).

That link is now invalid. Did you mean the `type DeviceMode string` definition? I doubt that this has that effect, because the Go type system cannot list all valid strings. `DeviceMode("foobar")` is valid in Go.

Let's not speculate, let's test it! (motto of the day, eh, week/month/year/career ...) :grin:

Reply:
> That link is now invalid.

One tip regarding that: what I do for long-running PRs with multiple links and rebases is to tag the current branch before rebasing and pushing that tag. That way, links that point to commits remain valid.

Reply:
> I doubt that this has that effect, because the Go type system cannot list all valid strings. DeviceMode("foobar") is valid in Go.

I was wrong, it does have an effect:
```
$ kubectl apply -f operator/examples/pmem-csi.intel.com_v1alpha1_deployment_cr.yaml 
The Deployment "example-deployment" is invalid: spec.deviceMode: Unsupported value: "foobar": supported values: "lvm", "direct"
```

Excellent. We still need an automated test, though.

Reply:
> > That link is now invalid.
> 
> One tip regarding that: what I do for long-running PRs with multiple links and rebases is to tag the current branch before rebasing and pushing that tag. That way, links that point to commits remain valid.

Thanks @pohly for the useful tip.